File: tessoku/A12.py
from __future__ import annotations

N, K = map(int, input().split())
A = [int(i) for i in input().split()]

def check(x: int, target: int, a: list[int]) -> bool:
    """ targetはx秒後の合計枚数s以下か
    """
    s = 0
    # a_i: 何秒ごとに1枚印刷するか (sec/枚)
    # x秒までにa_iは何枚印刷してるかを求めてる
    # 求めた合計の枚数がtarget以上かどうかを判定して返す
    # x秒後にtarget枚数分印刷できるか
    # targetがKに相当して、target以下ならそもそも印刷できないからx秒後では無理
    for a_i in a:
        s += x // a_i
    return s >= target

def binary_search(target: int, a: list[int]) -> int:
    left = 0
    right = 10 ** 9
    while left < right:
        # left + (right - left) / 2
        mid = (left + right) // 2
        # 答えはx秒以下か？
        # Trueなら中央寄りも左にあるってことだから右端の範囲を狭める
        # 反対に、Falseなら中央よりも右側にあるってことだから左端の範囲を狭める
        if check(x=mid, target=target, a=a):
            right = mid
        else:
            left = mid + 1
    return left


print(binary_search(target=K, a=A))

File: tessoku/A12.cpp
#include <iostream>
#include <vector>

/**
 * 答えがx以下かを判定して、Yesならtrueを返す
 *
 * @param x 判定する値. 何秒後か [s]
 * @param v 配列. 各要素はプリンターが１枚印刷するのにかかる時間[s/枚]
 * @param target 総和がtarget以上かどうか
 *
 * @return x以下ならtrue, それ以外はfalse
 */
bool check(long long x, std::vector<long long> &v, long long target) {
  long long n = v.size();
  long long sum = 0;
  for (long long i = 0; i < n; ++i) {
    sum += x / v[i];
  }
  if (sum >= target) {
    return true;
  } else {
    return false;
  }
}

int main() {
  using namespace std;
  long long n, k;
  cin >> n >> k;
  vector<long long> a(n);
  for (long long i = 0; i < n; ++i) {
    cin >> a[i];
  }

  // 何秒後にk枚以上の印刷がされるかを二分探索
  long long l = 0, r = 1'000'000'000;
  while (l < r) {
    long long mid = (l + r) / 2;
    bool ok = check(mid, a, k);
    if (ok) {
      r = mid; // 答えがmid以下なら、さらに小さい値
    } else {
      l = mid + 1; // 答えがmidより大きいなら、mid+1以上
    }
  }
  cout << l << endl; // lが最小の値
  return 0;
}